src/utils/DependencyUtils.ts

```diff
@@ -11,3 +29,71 @@ export const splitToNameAndVersion = (dependencyString: string): [string, string
     const fullName = parts.join('-');
     return [fullName, version];
 };
+
+/**
+ * - When installing a regular mods, always install the latest versions of dependencies.
+ * - When installing a mod pack we usually want to ensure compatibility by installing the exact
+ *      versions of dependencies, even if it means upgrading/downgrading already installed mods.
+ *   - If, however, modpackDependencyStrategy is set to USE_LATEST_VERSION, we install the latest
+ *      versions of dependencies for mod packs too.
+ * - When installing latest versions of dependencies, already installed mods are filtered out.
+ * - The order of the input combos array matters, combos earlier on the array get their dependencies'
+ *      versions prioritised over later ones.
+ */
+export async function getFullDependencyList(
+    combos: ThunderstoreCombo[],
+    game: Game,
+    installedMods: ManifestV2[],
+    modpackDependencyStrategy: ModpackDependencyStrategy
+) {
+    const originalCombos = [...combos];
+    for (const combo of originalCombos) {
+        const isModpack = combo.getMod().getCategories().some(value => value === "Modpacks");
+        const dependencyVersionStrategy = isModpack ? modpackDependencyStrategy : ModpackDependencyStrategy.USE_LATEST_VERSION;
+
+        let dependencies: ThunderstoreCombo[] = [];
+        await buildDependencySet(combo.getVersion(), game, dependencies, dependencyVersionStrategy);
+
+        combos.push(...dependencies.filter(dependency => {
+            const isNotAlreadyFound = !combos.some(alreadyAdded => alreadyAdded.getMod().getFullName() === dependency.getMod().getFullName());
```

Comment:
The original implementation of "install single package" downloader doesn't actually do this check against an array that contains the "main" package itself. So the subtle difference here is that in the original implementation, in a case of circular dependencies, the result might contain the same package twice. Since I consider this a very rare/speculative case, and the change basically a bug fix, I think the change should be included even if it doesn't match the original implementation.